src/headerssync.cpp

```diff
@@ -0,0 +1,318 @@
+    m_download_state = State::INITIAL_DOWNLOAD;
+
+    m_chain_start = chain_start;
+    m_minimum_required_work = minimum_required_work;
+    m_current_chain_work = chain_start->nChainWork;
+    m_current_height = chain_start->nHeight;
+    m_chain_start_locator = std::move(chain_start_locator);
+
+    m_last_header_received = m_chain_start->GetBlockHeader();
+
+    // Estimate the number of blocks that could possibly exist on the peer's
+    // chain *right now* using 6 blocks/second (fastest blockrate given the MTP
+    // rule) times the number of seconds from the last allowed block until
+    // today. This serves as a memory bound on how many commitments we might
+    // store from this peer, and we can safely give up syncing if the peer
+    // exceeds this bound, because it's not possible for a consensus-valid
+    // chain to be longer than this (at the current time -- in the future we
+    // could try again, if necessary, to sync a longer chain).
+    m_max_commitments = 6*(GetAdjustedTime() - chain_start->GetMedianTimePast() + MAX_FUTURE_BLOCK_TIME) / HEADER_COMMITMENT_FREQUENCY;
+
+    if (!ValidateAndStoreHeadersCommitments(initial_headers)) {
+        return std::nullopt;
+    }
+    return MakeNextHeadersRequest();
+}
+
+/** Process the next batch of headers received from our peer.
+ *  Validate and store commitments, and compare total chainwork to our target to
+ *  see if we can switch to REDOWNLOAD mode.  */
+HeadersSyncState::ProcessingResult HeadersSyncState::ProcessNextHeaders(const
+        std::vector<CBlockHeader>& headers, const bool full_headers_message)
+{
+    ProcessingResult ret;
+
+    Assume(!headers.empty());
+    if (headers.empty()) return ret;
+
+    Assume(m_download_state != State::FINAL);
+    if (m_download_state == State::FINAL) return ret;
+
+    if (m_download_state == State::INITIAL_DOWNLOAD) {
+        // During INITIAL_DOWNLOAD, we minimally validate block headers and
+        // occasionally add commitments to them, until we reach our work
+        // threshold.
+        if (!ValidateAndStoreHeadersCommitments(headers)) {
+            // The headers didn't pass validation; give up on the sync.
+            return ret;
+        }
+        ret.success = true;
+        if (full_headers_message || m_download_state == State::REDOWNLOAD) {
+            // A full headers message means the peer may have more to give us;
+            // also if we just switched to REDOWNLOAD then we need to re-request
+            // headers from the beginning.
+            ret.locator = MakeNextHeadersRequest();
+            return ret;
+        } else {
+            // If we're in INITIAL_DOWNLOAD and we get a non-full headers
+            // message, then the peer's chain has ended and definitely doesn't
+            // have enough work, so we can stop our sync.
+            LogPrint(BCLog::NET, "Ignoring low-work chain (height=%u) from peer=%d\n", m_current_height, m_id);
```

Comment:
I'm still a bit unclear on the sequence of calls here. Why are there two checks for non-full messages with low-work?

Reply:
Just to make sure I understand your question -- I think you're wondering why this log message appears here as well as in `net_processing.cpp` at line 2473?

The issue is that in `net_processing` we won't bother even starting a headers sync using the logic in `headerssync.cpp` if we receive a headers message that (a) isn't full and (b) which connects to something we know in our block index and (c) has too little work, because we know in this case that the headers don't lead to a high work chain (if it did, the headers message would have been full, indicating the peer has more to give us).

However, in the case that we start a headers sync using this module, then `net_processing` will be receiving headers that don't connect to the block index, and those messages can only be properly processed by this logic which keeps track of where we expect the headers to connect. So in this logic we also are checking whether the sync has to end because our peer has no more headers to give us (ie the headers message we processed from them was not full, and the chain has too little work).

Reply:
That was indeed my question.

Reply:
Oh, `ValidateAndStoreHeadersCommitments` has the side-effect of changing `m_download_state` to `REDOWNLOAD` once minimum difficulty is reached.